streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java

```diff
@@ -424,23 +400,60 @@ public void evictWhile(final Supplier<Boolean> predicate,
         }
     }
 
+    @Override
+    public boolean hasKey(final K key) {
+        final Bytes serializedKey = Bytes.wrap(keySerde.serializer().serialize(changelogTopic, key));
+        return index.containsKey(serializedKey);
+    }
+
+    @Override
+    public ValueAndTimestamp<V> priorValueForBuffered(final K key) {
+        final Bytes serializedKey = Bytes.wrap(keySerde.serializer().serialize(changelogTopic, key));
+        final BufferKey bufferKey = index.get(serializedKey);
+        if (bufferKey == null) {
+            throw new NoSuchElementException("Key [" + key + "] is not in the buffer.");
+        } else {
+            final BufferValue bufferValue = sortedMap.get(bufferKey);
+            final byte[] bytes = bufferValue.priorValue();
+            final V deserialize = valueSerde.innerSerde().deserializer().deserialize(changelogTopic + "-prior", bytes);
+            // it's unfortunately not possible to know this, unless we materialize the suppressed result, since our only
+            // knowledge of the prior value is what the upstream processor sends us as the "old value" when we first
+            // buffer something.
+            return ValueAndTimestamp.make(deserialize, RecordQueue.UNKNOWN);
```

Comment:
Hm. How important is it to get this timestamp right? It seems like it might be necessary to store the old timestamp in addition to the old value...

Reply:
Offhand, I'd say not that important.

Many of the other valuegetters in the KTable API also default to "unknown" == -1 at the moment.

It means we have weaker-than-desired time semantics, but since we have no defined semantics at this point, it's not actually violating anything.

The solution to this is quite large in scope, so I'm leaning toward keeping this as-is for now and fixing it in a later version.